src/lang/lang_values.ml

```diff
@@ -639,6 +644,7 @@ let rec check ?(print_toplevel = false) ~level ~env e =
     e.t >: mk (T.Arrow (proto_t, body.t))
   in
   match e.term with
+    | Abstract (n, _) -> e.t >: mkg (T.Abstract n)
```

Comment:
This is the only part of these changes that I don't like. Ideally, abstract types should be decoupled from opaque values. Here are a couple of thoughts in this topic:
* Being a scripted language, I don't think that the language will ever support the same paradigm as in OCaml when you can mask a type's concrete value, making it abstract
* Thus, abstract types for us will most likely only ever be used for opaque values, like a binding's configuration element and etc.

Reply:
If these are correct assumptions then the implementation is also probably correct as well.